kad-dht/README.md

```diff
@@ -0,0 +1,288 @@
+# libp2p Kademlia DHT specification
+
+The Kademlia Distributed Hash Table (DHT) subsystem in libp2p is a DHT
+implementation largely based on the Kademlia [0] whitepaper, augmented with
+notions from S/Kademlia [1], Coral [2] and mainlineDHT \[3\].
+
+This specification assumes the reader has prior knowledge of those systems. So
+rather than explaining DHT mechanics from scratch, we focus on differential
+areas:
+
+1. Specialisations and peculiarities of the libp2p implementation.
+2. Actual wire messages.
+3. Other algorithmic or non-standard behaviours worth pointing out.
+
+For everything else that isn't explicitly stated herein, it is safe to assume
+behaviour similar to Kademlia-based libraries.
+
+Code snippets use a Go-like syntax.
+
+## Authors
+
+* Protocol Labs.
+
+## Editors
+
+* [Raúl Kripalani](https://github.com/raulk)
+* [John Hiesey](https://github.com/jhiesey)
+
+## Distance function (dXOR)
+
+The libp2p Kad DHT uses the **XOR distance metric** as defined in the original
+Kademlia paper [0]. Peer IDs are normalised through the SHA256 hash function.
+
+For recap, `dXOR(sha256(id1), sha256(id2))` is the number of common leftmost
+bits between SHA256 of each peer IDs. The `dXOR` between us and a peer X
+designates the bucket index that peer X will take up in the Kademlia routing
+table.
+
+## Kademlia routing table
+
+The data structure backing this system is a k-bucket routing table, closely
+following the design outlined in the Kademlia paper [0]. The default value for
+`k` is 20, and the maximum bucket count matches the size of the SHA256 function,
+i.e. 256 buckets.
+
+The routing table is unfolded lazily, starting with a single bucket a position 0
+(representing the most distant peers), and splitting it subsequently as closer
+peers are found, and the capacity of the nearmost bucket is exceeded.
+
+## Alpha concurrency factor (α)
+
+The concurrency of node and value lookups are limited by parameter `α`, with a
+default value of 3. This implies that each lookup process can perform no more
+than 3 inflight requests, at any given time.
+
+## Record keys
+
+Records in the DHT are keyed by CID [4]. There are intentions to move to
+multihash [5] keys in the near future, as certain CID components like the
+multicodec are redundant.
+
+## Interfaces
+
+The libp2p Kad DHT implementation satisfies the routing interfaces:
```

Comment:
Is that really relevant to a spec? That looks pretty specific to Go to me, or to a specific set of programming languages that are capable of fulfilling it. In particular the Rust code has no interest in following these interfaces.

Reply:
I had second thoughts when dumping the interface here, as it should be non-normative as you say. However, it helps bind things together, as it specifies the public calls supported by this component along with its inputs and outputs, i.e. the expected public surface of this component. Also, each exposed behaviour is specified at some point in the doc.

IMHO, we do need to capture an abstract interface outline, and I used Go nomenclature and copied our existing one. I'm open to changing this.

Reply:
We have separate repos defining at least two of these interfaces abstractly; see https://github.com/libp2p/interface-peer-routing https://github.com/libp2p/interface-content-routing

Admittedly this isn't required for interoperability, but I think we should at least suggest relevant interfaces for the DHT's public API.

Reply:
I think shoehorning the DHT node implementation into these external interfaces is causing it to take on unnecessary complexity. Using the DHT for routing (to fit the interface) should be trivial to provide with a type that wraps the DHT node. This will free up some very bizarre methods and requirements that currently exist directly on the node implementation. Perhaps the spec should say that "The DHT node implementation may implement or provide the required features to implement the following interfaces:" or something to that effect.

Reply:
I'm inclined to agree, yes

Reply:
https://github.com/libp2p/specs/pull/108/commits/f05ee26ed4636a23f00ed86eeddf963ebb4de689 removes the DHT interfaces, capturing its structure and naming in the structure of the description itself. The `Validator` interface is kept as is, though the requirement of strictly implementing the interface is loosened. 

Let me know in case you have any objections.